File: src/py_examples/py_examples/migrate_system_state.py
#!/usr/bin/env python3
"""
System State Migrator - Synchronous Python version
"""

import sys
import time
import rclpy
from rclpy.node import Node
import aimdk_msgs.srv as aimdk_srv
import aimdk_msgs.msg as aimdk_msg


class SyncSystemStateMigrator(Node):
    """Synchronous version of system state migrator"""

    def __init__(self, target_state: str):
        super().__init__('sync_system_state_migrator')
        self.target_state = target_state

        # Create service clients
        self.migrate_client = self.create_client(
            aimdk_srv.MigrateSystemState,
            '/aimdk_5Fmsgs/srv/MigrateSystemState'
        )

        self.get_state_client = self.create_client(
            aimdk_srv.GetSystemState,
            '/aimdk_5Fmsgs/srv/GetSystemState'
        )

        self.get_logger().info(f'Target state: {self.target_state}')

    def wait_for_services(self, timeout_sec: float = 30.0) -> bool:
        """Wait for services to become available"""
        self.get_logger().info('Waiting for services...')

        start_time = time.time()
        while time.time() - start_time < timeout_sec and rclpy.ok():
            if (self.migrate_client.wait_for_service(timeout_sec=1.0) and
                    self.get_state_client.wait_for_service(timeout_sec=1.0)):
                self.get_logger().info('Services are available')
                return True

            self.get_logger().info('Services not available, waiting...')
            time.sleep(0.5)

        self.get_logger().error(
            f'Timeout waiting for services after {timeout_sec} seconds')
        return False

    def migrate_state_sync(self) -> bool:
        """Send migration request synchronously"""
        request = aimdk_srv.MigrateSystemState.Request()
        now = self.get_clock().now()
        request.header.header.stamp.sec = now.nanoseconds // 1_000_000_000
        request.header.header.stamp.nanosec = now.nanoseconds % 1_000_000_000
        request.state = self.target_state

        self.get_logger().info(
            f'Sending migration request to {self.target_state}...')

        try:
            # Use call() for synchronous call (requires ROS2 Foxy or newer)
            # For older versions, use call_async() with spin_until_future_complete
            future = self.migrate_client.call_async(request)

            # Spin until future completes
            start_time = time.time()
            while rclpy.ok() and time.time() - start_time < 10.0:
                rclpy.spin_once(self, timeout_sec=0.1)
                if future.done():
                    response = future.result()
                    if response.header.status.value == aimdk_msg.CommonState.SUCCESS:
                        self.get_logger().info('Migration request accepted')
                        return True
                    else:
                        self.get_logger().error(
                            f'Migration failed: {response.header.message}'
                        )
                        return False

            self.get_logger().error('Timeout waiting for migration response')
            return False

        except Exception as e:
            self.get_logger().error(f'Migration service call failed: {e}')
            return False

    def get_current_state_sync(self):
        """Get current system state synchronously"""
        request = aimdk_srv.GetSystemState.Request()
        now = self.get_clock().now()
        request.header.header.stamp.sec = now.nanoseconds // 1_000_000_000
        request.header.header.stamp.nanosec = now.nanoseconds % 1_000_000_000

        try:
            future = self.get_state_client.call_async(request)

            # Spin until future completes
            start_time = time.time()
            while rclpy.ok() and time.time() - start_time < 5.0:
                rclpy.spin_once(self, timeout_sec=0.1)
                if future.done():
                    return future.result()

            self.get_logger().error('Timeout getting system state')
            return None

        except Exception as e:
            self.get_logger().error(f'GetSystemState service call failed: {e}')
            return None

    def monitor_migration_sync(self, max_checks: int = 300) -> bool:
        """Monitor migration progress synchronously"""
        self.get_logger().info('Monitoring migration progress...')

        check_count = 0
        start_time = time.time()

        while check_count < max_checks and rclpy.ok():
            # Wait between checks
            time.sleep(1.0)

            # Check current state
            response = self.get_current_state_sync()
            if response is None:
                self.get_logger().warning('Failed to get system state, will retry...')
                continue

            check_count += 1

            # Check if migration is complete
            current_state_lower = response.cur_state.lower()
            target_state_lower = self.target_state.lower()

            state_match = (current_state_lower == target_state_lower)
            status_match = (response.curr_status.value ==
                            aimdk_msg.SystemStatus.IN_READY)

            self.get_logger().info(
                f'Check {check_count}: State="{response.cur_state}" '
                f'(match={state_match}), Status={response.curr_status.value} (match={status_match})'
            )

            if state_match and status_match:
                self.get_logger().info(
                    f'Migration to {self.target_state} completed successfully!')
                return True

            # Check timeout
            if time.time() - start_time > 300.0:  # 5 minutes total timeout
                self.get_logger().error('Migration timeout after 5 minutes')
                return False

        self.get_logger().error(f'Migration failed after {max_checks} checks')
        return False

    def run_migration(self) -> bool:
        """Run complete migration process"""
        # Step 1: Wait for services
        if not self.wait_for_services():
            return False

        # Step 2: Send migration request
        if not self.migrate_state_sync():
            return False

        # Step 3: Monitor migration progress
        return self.monitor_migration_sync()


def main():
    """
    Main function
    """
    # Check command line arguments
    if len(sys.argv) != 2:
        print(f'Usage: {sys.argv[0]} <target_state>')
        print('\nAvailable target states: Ready, Develop_Nav, Develop_Audio_Linux, Develop_Audio_ROS, Develop_MC')
        return 1

    # Get target state from command line argument
    target_state = sys.argv[1]

    # Validate target state
    if not target_state:
        print('Error: Target state cannot be empty', file=sys.stderr)
        return 1

    # Initialize ROS2
    rclpy.init()

    result = 1
    node = None

    try:
        # Create migrator node
        node = SyncSystemStateMigrator(target_state)

        # Run migration
        if node.run_migration():
            print(f'\nMigration to {target_state} completed successfully!')
            result = 0
        else:
            print(f'\nMigration to {target_state} failed')
            result = 1

    except KeyboardInterrupt:
        print('\nProgram interrupted by user')
        result = 1
    except Exception as e:
        print(f'Error: {e}', file=sys.stderr)
        result = 1
    finally:
        # Cleanup
        if node:
            node.destroy_node()
        rclpy.shutdown()

    return result


if __name__ == '__main__':
    sys.exit(main())

File: src/examples/src/utils/migrate_system_state.cpp
#include <algorithm>
#include <chrono>
#include <iostream>
#include <memory>
#include <string>

#include "aimdk_msgs/srv/get_system_state.hpp"
#include "aimdk_msgs/srv/migrate_system_state.hpp"
#include "rclcpp/rclcpp.hpp"

using namespace std::chrono_literals;

/**
 * @class SystemStateMigrator
 * @brief ROS2 client for migrating system state to any target state
 */
class SystemStateMigrator : public rclcpp::Node {
public:
  /**
   * @brief Constructor for SystemStateMigrator
   * @param target_state The target system state to migrate to
   */
  SystemStateMigrator(const std::string &target_state)
      : Node("system_state_migrator"), target_state_(target_state) {
    // Create service clients
    migrate_client_ = this->create_client<aimdk_msgs::srv::MigrateSystemState>(
        "/aimdk_5Fmsgs/srv/MigrateSystemState");

    get_state_client_ = this->create_client<aimdk_msgs::srv::GetSystemState>(
        "/aimdk_5Fmsgs/srv/GetSystemState");

    // Log target state
    RCLCPP_INFO(this->get_logger(), "Target state: %s", target_state_.c_str());

    // Start the migration process
    start_migration_process();
  }

private:
  /**
   * @brief Start the migration process
   */
  void start_migration_process() {
    // Create a timer to check service availability
    check_timer_ = this->create_wall_timer(
        1s, std::bind(&SystemStateMigrator::check_services_and_start, this));
  }

  /**
   * @brief Check if services are available and start migration
   */
  void check_services_and_start() {
    // Check if both services are available
    if (migrate_client_->service_is_ready() &&
        get_state_client_->service_is_ready()) {
      // Services are available, cancel timer and start migration
      check_timer_->cancel();
      RCLCPP_INFO(this->get_logger(),
                  "Both services are available, starting migration to %s...",
                  target_state_.c_str());
      migrate_to_target_state();
    } else {
      RCLCPP_INFO(this->get_logger(), "Waiting for services...");
    }
  }

  /**
   * @brief Migrate system state to target state
   */
  void migrate_to_target_state() {
    // Create migration request
    auto request =
        std::make_shared<aimdk_msgs::srv::MigrateSystemState::Request>();

    // Set request timestamp
    auto now = this->now();
    request->header.header.stamp.sec = now.seconds();
    request->header.header.stamp.nanosec = now.nanoseconds() % 1000000000;

    // Set target state from parameter
    request->state = target_state_;

    RCLCPP_INFO(this->get_logger(), "Sending migration request to %s state...",
                target_state_.c_str());

    // Send migration request
    auto future = migrate_client_->async_send_request(
        request, std::bind(&SystemStateMigrator::migration_response_callback,
                           this, std::placeholders::_1));
  }

  /**
   * @brief Callback for migration response
   * @param future Future object containing migration response
   */
  void migration_response_callback(
      rclcpp::Client<aimdk_msgs::srv::MigrateSystemState>::SharedFuture
          future) {
    try {
      // Get migration response
      auto response = future.get();

      // Check migration response
      if (response->header.status.value ==
          aimdk_msgs::msg::CommonState::SUCCESS) {
        RCLCPP_INFO(this->get_logger(),
                    "Migration request accepted, starting to monitor state...");

        // Start monitoring system state
        start_state_monitoring();

      } else {
        RCLCPP_ERROR(this->get_logger(),
                     "Migration request failed with status: %d, message: %s",
                     response->header.status.value,
                     response->header.message.c_str());
        rclcpp::shutdown();
      }

    } catch (const std::exception &e) {
      RCLCPP_ERROR(this->get_logger(), "Migration service call failed: %s",
                   e.what());
      rclcpp::shutdown();
    }
  }

  /**
   * @brief Start monitoring system state
   */
  void start_state_monitoring() {
    RCLCPP_INFO(this->get_logger(), "Starting to monitor system state...");

    // Create a timer to periodically check system state
    monitor_timer_ = this->create_wall_timer(
        1s, std::bind(&SystemStateMigrator::check_system_state, this));
  }

  /**
   * @brief Check current system state
   */
  void check_system_state() {
    // Create get state request
    auto request = std::make_shared<aimdk_msgs::srv::GetSystemState::Request>();

    // Set request timestamp
    auto now = this->now();
    request->header.header.stamp.sec = now.seconds();
    request->header.header.stamp.nanosec = now.nanoseconds() % 1000000000;

    // Send get state request
    auto future = get_state_client_->async_send_request(
        request, std::bind(&SystemStateMigrator::state_response_callback, this,
                           std::placeholders::_1));
  }

  /**
   * @brief Callback for get state response
   * @param future Future object containing state response
   */
  void state_response_callback(
      rclcpp::Client<aimdk_msgs::srv::GetSystemState>::SharedFuture future) {
    try {
      // Get state response
      auto response = future.get();

      // Log current state for debugging
      RCLCPP_INFO(this->get_logger(), "Current State: %s, System Status: %u",
                  response->cur_state.c_str(), response->curr_status.value);

      // Check if migration is complete
      if (is_migration_complete(response)) {
        RCLCPP_INFO(this->get_logger(),
                    "Migration to %s completed successfully!",
                    target_state_.c_str());
        monitor_timer_->cancel();
        rclcpp::shutdown();
      } else {
        RCLCPP_INFO(this->get_logger(),
                    "Migration in progress, will check again in 1 second...");
      }

    } catch (const std::exception &e) {
      RCLCPP_ERROR(this->get_logger(), "GetSystemState service call failed: %s",
                   e.what());
    }
  }

  /**
   * @brief Check if migration to target state is complete
   * @param response Response from GetSystemState service
   * @return true if migration is complete, false otherwise
   */
  bool is_migration_complete(
      const aimdk_msgs::srv::GetSystemState::Response::SharedPtr response) {
    // Convert current state and target state to lowercase for case-insensitive
    // comparison
    std::string current_state_lower = response->cur_state;
    std::transform(current_state_lower.begin(), current_state_lower.end(),
                   current_state_lower.begin(), ::tolower);

    std::string target_state_lower = target_state_;
    std::transform(target_state_lower.begin(), target_state_lower.end(),
                   target_state_lower.begin(), ::tolower);

    // Check conditions:
    // 1. Current state equals target state (case-insensitive)
    // 2. System status equals 1 (IN_READY)
    bool state_match = (current_state_lower == target_state_lower);
    bool status_match = (response->curr_status.value ==
                         aimdk_msgs::msg::SystemStatus::IN_READY);

    if (state_match && status_match) {
      return true;
    }

    // Log detailed information if not complete
    if (!state_match) {
      RCLCPP_INFO(this->get_logger(),
                  "State mismatch: current='%s', expected='%s'",
                  response->cur_state.c_str(), target_state_.c_str());
    }
    if (!status_match) {
      RCLCPP_INFO(this->get_logger(),
                  "Status mismatch: current= %u, expected=1 (IN_READY)",
                  response->curr_status.value);
    }
    if (state_match && !state_match) {
      RCLCPP_ERROR(this->get_logger(), "State Migrate Fail");
      exit(0);
    }

    return false;
  }

  // Target state to migrate to
  std::string target_state_;

  // Service clients
  rclcpp::Client<aimdk_msgs::srv::MigrateSystemState>::SharedPtr
      migrate_client_;
  rclcpp::Client<aimdk_msgs::srv::GetSystemState>::SharedPtr get_state_client_;

  // Timers
  rclcpp::TimerBase::SharedPtr check_timer_;
  rclcpp::TimerBase::SharedPtr monitor_timer_;
};

/**
 * @brief Print usage information with detailed mode descriptions
 */
void print_usage(const char *program_name) {
  std::cout << "Usage: " << program_name << " <target_state>" << std::endl;
  std::cout << std::endl;
  std::cout << "Available target states:" << std::endl;
  std::cout << "  " << program_name << " Ready" << std::endl;
  std::cout << "  " << program_name << " Develop_Nav" << std::endl;
  std::cout << "  " << program_name << " Develop_Audio_Linux" << std::endl;
  std::cout << "  " << program_name << " Develop_Audio_ROS" << std::endl;
  std::cout << "  " << program_name << " Develop_MC" << std::endl;
  std::cout << std::endl;

  std::cout << "Mode Descriptions:" << std::endl;
  std::cout << "==================" << std::endl;

  // Ready mode description
  std::cout << "1. Ready - System Default Mode" << std::endl;
  std::cout
      << "   • Description: System enters this mode by default after startup"
      << std::endl;
  std::cout << "   • Purpose: Normal operation mode, ready for general tasks"
            << std::endl;
  std::cout
      << "   • Use case: Return to normal operation after development work"
      << std::endl;
  std::cout << std::endl;

  // Develop_Nav mode description
  std::cout << "2. Develop_Nav - Navigation Development Mode" << std::endl;
  std::cout
      << "   • Description: Special mode for navigation system development"
      << std::endl;
  std::cout << "   • Purpose: Enables navigation-related debugging and testing"
            << std::endl;
  std::cout << "   • Use case: When developing or testing navigation "
               "algorithms, SLAM, path planning"
            << std::endl;
  std::cout << "   • Features: May provide access to raw sensor data, "
               "navigation debug topics"
            << std::endl;
  std::cout << std::endl;

  // Develop_Audio_Linux mode description
  std::cout << "3. Develop_Audio_Linux - System-Level Audio Development Mode"
            << std::endl;
  std::cout << "   • Description: Mode for low-level audio system development"
            << std::endl;
  std::cout << "   • Purpose: Access to system audio streams at Linux level"
            << std::endl;
  std::cout << "   • Use case: When developing audio drivers, audio processing "
               "at system level"
            << std::endl;
  std::cout << "   • Features: Direct access to audio hardware, "
               "ALSA/PulseAudio interfaces"
            << std::endl;
  std::cout << std::endl;

  // Develop_Audio_ROS mode description
  std::cout << "4. Develop_Audio_ROS - ROS Audio Development Mode" << std::endl;
  std::cout
      << "   • Description: Mode for ROS-based audio application development"
      << std::endl;
  std::cout << "   • Purpose: Access to audio data through ROS topics"
            << std::endl;
  std::cout << "   • Use case: When developing ROS nodes that process audio, "
               "speech recognition"
            << std::endl;
  std::cout << "   • Features: Audio data published as ROS topics, ROS message "
               "interfaces"
            << std::endl;
  std::cout << std::endl;

  // Develop_MC mode description
  std::cout << "5. Develop_MC - Motion Control Development Mode" << std::endl;
  std::cout << "   • Description: Mode for motion control system development"
            << std::endl;
  std::cout
      << "   • Purpose: Enables direct control and testing of motion systems"
      << std::endl;
  std::cout << "   • Use case: When developing motor control algorithms, "
               "testing motion hardware"
            << std::endl;
  std::cout << "   • Features: Low-level motor control access, motion system "
               "debugging tools"
            << std::endl;
  std::cout << std::endl;

  std::cout << "Program Workflow:" << std::endl;
  std::cout << "=================" << std::endl;
  std::cout
      << "1. Calls MigrateSystemState service with the specified target state"
      << std::endl;
  std::cout << "2. Continuously monitors GetSystemState service every second"
            << std::endl;
  std::cout << "3. Exits when both conditions are met:" << std::endl;
  std::cout << "   • Current State equals target state (case-insensitive)"
            << std::endl;
  std::cout << "   • System Status equals 1 (IN_READY)" << std::endl;
  std::cout << std::endl;
  std::cout << "Note: The program will automatically exit when migration is "
               "complete or on error."
            << std::endl;
}

/**
 * @brief Validate if the target state is a known development mode
 * @param state The state to validate
 * @return true if state is valid, false otherwise
 */
bool is_valid_state(const std::string &state) {
  // List of known valid states
  const std::vector<std::string> valid_states = {
      "Ready", "Develop_Nav", "Develop_Audio_Linux", "Develop_Audio_ROS",
      "Develop_MC"};

  // Convert input to lowercase for case-insensitive comparison
  std::string state_lower = state;
  std::transform(state_lower.begin(), state_lower.end(), state_lower.begin(),
                 ::tolower);

  // Check if state is in the list of valid states
  for (const auto &valid_state : valid_states) {
    std::string valid_state_lower = valid_state;
    std::transform(valid_state_lower.begin(), valid_state_lower.end(),
                   valid_state_lower.begin(), ::tolower);

    if (state_lower == valid_state_lower) {
      return true;
    }
  }

  return false;
}

/**
 * @brief Main function
 * @param argc Argument count
 * @param argv Argument vector
 * @return Exit code
 */
int main(int argc, char **argv) {
  // Check command line arguments
  if (argc != 2) {
    print_usage(argv[0]);
    return 1;
  }

  // Get target state from command line argument
  std::string target_state = argv[1];

  // Validate target state
  if (target_state.empty()) {
    std::cerr << "Error: Target state cannot be empty" << std::endl;
    print_usage(argv[0]);
    return 1;
  }

  // Optional: Validate against known states
  if (!is_valid_state(target_state)) {
    std::cout << "Warning: '" << target_state
              << "' is not in the list of known states." << std::endl;
    std::cout << "The program will still attempt to migrate to this state."
              << std::endl;
    std::cout << "Continue? (y/n): ";

    std::string response;
    std::getline(std::cin, response);

    if (response != "y" && response != "Y") {
      std::cout << "Operation cancelled." << std::endl;
      return 0;
    }
  }

  // Initialize ROS2
  rclcpp::init(argc, argv);

  // Create migrator node with target state
  auto node = std::make_shared<SystemStateMigrator>(target_state);

  // Create single-threaded executor
  auto executor = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();

  // Add node to executor
  executor->add_node(node);

  // Run executor (blocking call)
  executor->spin();

  // Cleanup
  executor->remove_node(node);
  rclcpp::shutdown();

  return 0;
}